# Interactive MA plot for data exploration
#
# MA data and stats generated by:  
#   1) TopHat2 alignment  
#   2) feature count with HTSeq  
#   3) feature statistics with EdgeR (performed below)
#
# Note that this file should be in the same directory as the MAplot_shiny_app
# directory containing the the app.R file.  Also the meta-data 
# (MAplot_shiny_meta_data.csv) and sample files (sampleX.txt) should be in the 
# same directory as this file.  That is:
#    ~/edgeR_MAplot_shiny_template.R
#    ~/MAplot_shiny_meta_data.csv
#    ~/sample1.txt
#    ~/sample2.txt
#    ~/sample3.txt
#    ~/sample4.txt
#
#    ~/MAplot_shiny_app/app.R
#
# Note that MA plot can be resized to default view by removing all values 
# for the x- and y-axis.  Also note, non-integer numbers can be manually 
# entered for finer resolution.

# packages
library(shiny)
library(edgeR)
library(ggplot2)
library(magrittr)

### create dataframe with feature IDs, M values, A values using edgeR package
  # read in metadata
  targets <- read.csv(file = "MAplot_shiny_meta_data.csv", stringsAsFactors = FALSE)

  # read in sample counts and relevel factors with respect to controls
  raw.counts <- readDGE(files = targets$sample_file, columns=c(1,2), group = as.factor(targets$condition)) 
  raw.counts$samples$group <- relevel(raw.counts$samples$group, ref = "condition1")

  # remove rows added by TopHat2 concerning mapping statistics (ex. __too_low_aQual);
  drop <- grepl(pattern = "__", rownames(raw.counts$counts))
  raw.counts$counts <- raw.counts$counts[!drop, ]

  # threshold out features that have less than a sum of 2 reads per million 
  # across all samples
  keep <- rowSums(cpm(raw.counts)) >= 2
  filtered.counts <- raw.counts[keep, ]

  # dataframe of log2(fold-change) and average log2(copies-per-million) returned 
  # by plotSmear()
  MA <- filtered.counts
  MA %<>% plotSmear(plot.it = FALSE) %>% do.call(rbind.data.frame, .) %>% t() %>% as.data.frame()
  MA <- MA[, 1:2]
  colnames(MA) <- c("av.log2cpm", "log2FC")
  rownames(MA) <- rownames(filtered.counts$counts)

### create column of significantly regulated genes to be highlighted in MAplot
### using edgeR

  # reset library size
  filtered.counts$samples$lib.size <- colSums(filtered.counts$counts)

  # normalize libraries with trimmed means of M-value method
  filtered.counts %<>% calcNormFactors(method = "TMM")

  # estimate dispersion and return feature-wise test statistics; enter
  # control and experimental IDs for the 'pair' argument
  filtered.counts %<>% estimateDisp()
  exact.tests <- exactTest(filtered.counts, pair = c("condition1", "condition2"))

  # significantly regulated genes highlighted in red; 10% false discovery
  col <- exact.tests
  col %<>% decideTestsDGE(adjust.method = "BH", p = 0.1) %>% as.vector() 
  col %<>% replace(col == 0, "black") 
  col %<>% replace(col == -1, "red")
  col %<>% replace(col == 1, "red")
  MA$col <- col

### run app
  runApp("MAplot_shiny_app") 